firmware/STN1110.c

```diff
@@ -103,33 +148,79 @@ static bool _starts_with_hex(char *buf)
     return _parse_byte(first_char_str, &first_value, 16);
 }
 
-static void _log_CAN_tx_message(CANTxFrame * can_frame)
+
+static void _decode_protocol(char * buf)
 {
-    if (get_logging_level() < logging_level_trace)
+    if (strstr(buf, "SAE J1850 PWM") != 0) {
+        set_detected_protocol(obdii_protocol_j1850_pwm);
+        return;
+    } else if (strstr(buf, "SAE J1850 VPW") != 0) {
+        set_detected_protocol(obdii_protocol_j1850_vpw);
+        return;
+    } else if (strstr(buf, "ISO 9141-2") != 0) {
+        set_detected_protocol(obdii_protocol_9141_2);
+        return;
+    } else if (strstr(buf, "ISO 14230-4") != 0) {
+        set_detected_protocol(obdii_protocol_iso14230_4_kwp_5baud);
         return;
-
-    uint32_t CAN_id = can_frame->IDE == CAN_IDE_EXT ? can_frame->EID : can_frame->SID;
-    log_trace(LOG_PFX "CAN Tx ID(%i): ", CAN_id);
-    size_t i;
-    for (i = 0; i < can_frame->DLC; i++)
-    {
-        log_trace("%02X ", can_frame->data8[i]);
     }
-    log_trace("\r\n");
 }
-void _process_pid_response(char * buf)
+
+void _process_stn1110_response(char * buf)
 {
+    /* If our received data includes the AT command prompt,
+     * then skip past it.
+     */
+    if (buf[0] == '>')
+        buf++;
+
+    /* check if we got a protocol response */
+    if (strstr(buf, "AUTO, ") != 0) {
+        _decode_protocol(buf + 6);
+        return;
+    }
+
+    /* So optimistic */
+    enum STN1110_error stn1110_result = STN1110_ERROR_NONE;
+
+    bool got_obd2_response = false;
     if (strstr(buf, "STOPPED") != 0) {
-        log_info(LOG_PFX "stopped\r\n");
-        chThdSleepMilliseconds(OBDII_PID_POLL_DELAY);
-        set_pid_request_active(false);
+        log_info(LOG_PFX "Stopped\r\n");
+        stn1110_result = STN1110_ERROR_STOPPED;
+        /* When we get the STOPPED message it means we're
+         * asking for data too fast. Stretch out the poll delay.
+         */
+        stretch_pid_poll_delay();
     }
+
+    /* Handle a couple of known errors */
     else if (strstr(buf, "NO DATA") != 0) {
-        log_info(LOG_PFX "no data\r\n");
-        chThdSleepMilliseconds(OBDII_PID_POLL_DELAY);
-        set_pid_request_active(false);
+        log_info(LOG_PFX "No data\r\n");
+        stn1110_result = STN1110_ERROR_NO_DATA;
+    } else if (strstr(buf, "ERROR") !=0 && strstr(buf, "BUS") != 0) {
```

Comment:
So it can come back in orders like "ERROR BUS" or "BUS ERROR" or why this separation?  Comments?